update readme with fix from #1939

kellyrmilligan opened the pull request:
updating readme with comment from #1939 . works with latest `node-sass@4.5.2`

🎉

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -406,7 +406,7 @@ Then in `package.json`, add the following lines to `scripts`:
    "scripts": {
 +    "build-css": "node-sass src/ -o src/",
-+    "watch-css": "npm run build-css && node-sass src/ -o src/ --watch --recursive",
++    "watch-css": "npm run build-css && node-sass src/*.scss -o src/ --watch --recursive",
```
Does this need to be `src/**/*.scss`?
Also, what about `sass` files? Aren't both extensions supported?

kellyrmilligan replied:
👀

kellyrmilligan replied:
ok, updating. I coudn't find anything at first in the readme of `node-sass` but looking through the code of the cli, yeah normal globbing is supported. updating.  `.sass` files are supported, but I think for a readme this should suffice no?

Timer replied:
Most likely, yeah! I just want to make sure it's noted.

Thanks!

kellyrmilligan replied:
noted 👍

Timer commented:
Hey @kellyrmilligan, thanks for this PR!

Do you happen to know about that comment I mentioned above?

cr101 commented:
@kellyrmilligan Unfortunately, it doesn't work on Windows

> Error: ENOENT: no such file or directory, lstat 'C:\wf\my-app\src\**\*.scss'

kellyrmilligan commented:
What if you tried app\src\**\*.scss ?

kellyrmilligan commented:
i'm looking more into the code now. there are a few levels of abstraction around how the globbing works, so I may change this back to the original fix and explain in this PR why it works.

kellyrmilligan commented:
looks like the recommended solution from the issue won't let you watch files in subdirectories. the glob pattern in the cli gets expanded, and then node-sass doesn't treat it as a directory. on the other hand, if you use `src/*/**.scss`, it doesn't work as you would expect either.  from what I am seeing, the way they want it to be used is `src/`. this results in the `sass-graph` lib picking up all the correct files. The cli assumes you want all extensions, and its not configurable. 

After trying out a few different ways of calling the cli, on my setup at least, the looping happens when you change a file that is included in a lot of files, like say a `_settings.scss` file or something. 

I switched over to `node-sass-chokidar` and this issue just went away. I think longer term the node-sass cli should let you configure the extensions or use proper globbing patterns, as ATM its kind of broken IMO.

switching to `node-sass-chokidar` fixed this issue fully for me, so not sure what to do with this PR unless we want to switch the recommendation for sass to that package instead. 

thoughts @Timer and anyone else?

Timer commented:
Let's just switch the docs to node-sass-chokidar primarily.

gaearon commented:
Would it work if you quoted the glob? `node-sass "src/**/*.scss"`

kellyrmilligan commented:
@gaearon let me give it a shot.

kellyrmilligan commented:
nope, it takes that as the literal value and gives `Error: ENOENT: no such file or directory, lstat 'src/**/*.scss'`

kellyrmilligan commented:
I updated the readme again. I reverted it back to the way it was, and added a note under the step to install `node-sass`,  about `node-sass-chokidar`.

Timer commented:
We already have a note to switch to it about performance, but I think we should just recommend it first and adjust both notes to say why we use it.

I don't see a need to show docs using node-sass.

Timer commented:
Perfect! Thanks much.

Timer commented:
Just to confirm, new files aren't an issue anymore?

kellyrmilligan commented:
@Timer added note about new files back in after testing. please let me know of any other changes I just pushed an update to the notes as well.

gaearon commented on packages/react-scripts/template/README.md:
```diff
@@ -207,7 +207,7 @@ To configure the syntax highlighting in your favorite text editor, head to the [
 
 ## Displaying Lint Output in the Editor
 
->Note: this feature is available with `react-scripts@0.2.0` and higher.  
```
Whitespace is significant in markdown. Please leave those two spaces in, they are causing the line to break. They were intentional.

Timer commented:
/cc @michaelwayman

michaelwayman commented on packages/react-scripts/template/README.md:
````diff
@@ -442,27 +441,17 @@ Then we can change `start` and `build` scripts to include the CSS preprocessor c
 ```
 
-Now running `npm start` and `npm run build` also builds Sass files. Note that `node-sass` seems to have an [issue recognizing newly created files on some systems](https://github.com/sass/node-sass/issues/1891) so you might need to restart the watcher when you create a file until it’s resolved.
+Now running `npm start` and `npm run build` also builds Sass files. Note that `node-sass-chokidar` seems to have an issue recognizing newly created files on some systems so you might need to restart the watcher when you create a file until it’s resolved.
````
@Timer thanks for the tag, this is amazing :)

I would like to comment though that `node-sass-chokidar` does not have a problem recognizing new files.

gaearon replied:
hmm, doesn't this contradict:

>added note about new files back in after testing

@kellyrmilligan Can you confirm you still experience the issue with new files?

kellyrmilligan replied:
@michaelwayman in testing it seems to not recognize a new file being added for me.

kellyrmilligan replied:
what is a good way to verify it? I am testing on os x sierra.

michaelwayman replied:
ok, thanks @kellyrmilligan 
can you open an issue on `node-sass-chokidar` and describe steps to reproduce, as well as your system details. (e.g. OS, node, npm, and package versions) thanks so much and I will try to have this fixed ASAP

It works as expected for me so I need to try and duplicate your setup.

michaelwayman replied:
Well, what I did to test it was:
 1. Start a clean project and install node-sass-chokidar.
 1. run `node-sass-chokidar src/ -o src/ --watch --recursive`
 1. Create a new scss file `src/my_file.scss` with simple style `#yoyo {color:red;}` and save.
    - note that the file *HAS* to be in the directory that's being watched i.e. `src/`
 1. Check for the existence of `src/my_file.css`

that was all the "testing" I did but it works fine. So yeah, please describe your steps to reproduce and system info in an issue [here](https://github.com/michaelwayman/node-sass-chokidar/issues)

kellyrmilligan replied:
I realized I tried a `_test.scss` file, which is meant to be included in others. in testing a regular file you are correct and i've removed the note about that.

michaelwayman commented:
That's awesome @kellyrmilligan thanks :)

And yep, files starting with `_` do not get transpiled.

I think the `not recognizing new files` is a known bug with `node-sass`.
Since you already have the "**Why `node-sass-chokidar`?**" section, if you want to add that nsc fixes it that would be super 👍  lol

 <3 this PR BTW :)

kellyrmilligan commented:
@michaelwayman done!

heshamsalman commented:
@michaelwayman 👏 👏 👏

Timer commented on packages/react-scripts/template/README.md:
````diff
@@ -398,15 +398,14 @@ Following this rule often makes CSS preprocessors less useful, as features like
 ```diff
    "scripts": {
-+    "build-css": "node-sass src/ -o src/",
-+    "watch-css": "npm run build-css && node-sass src/ -o src/ --watch --recursive",
````
Why do we build before we start watching? Does the watcher not compile everything when started in watch mode?

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -430,8 +429,8 @@ Then we can change `start` and `build` scripts to include the CSS preprocessor c
-     "build-css": "node-sass src/ -o src/",
-     "watch-css": "npm run build-css && node-sass src/ -o src/ --watch --recursive",
+     "build-css": "node-sass-chokidar src/ -o src/",
+     "watch-css": "npm run build-css && node-sass-chokidar src/ -o src/ --watch --recursive",
```
Ditto.

michaelwayman replied:
without the `build-css` inside of `watch-css` there is no guarantee on the latest stylesheets being used..

so,
 1. If it's the first time running `npm run start` you won't have ANY CSS files yet
 1. If you made changes to a stylesheet while you weren't watching then `npm run start` those changes won't show until you modify the stylesheet again

kellyrmilligan replied:
just to comment on this, I do like for example how the `babel` cli when started in watch mode does compile one time before the watch mode kicks in. something to think about and would simplify the scripts config.

michaelwayman replied:
@kellyrmilligan do you want to open an issue on node-sass-chokidar requesting this change? I will make the change this weekend if you do.

Timer replied:
If you make the files compile when started in watch mode, please add a flag to disable it.
It's very annoying that babel does it because you can't say "wait for first build before starting app".

ajs139 replied:
@michaelwayman: would it be helpful to include node_modules, like so:
```
    "build-css": "node-sass-chokidar src/ -o src/ --include-path node_modules",
    "watch-css": "npm run build-css && node-sass-chokidar src/ -o src/ --watch --recursive --include-path node_modules",
```

kellyrmilligan replied:
@ajs139 we actually recently added a section about that. It gives advice more about importing without relative paths similar to the NODE PATH env variable fix. I always add my src directory as well.

kidqn replied:
Can we hide css compiled files in "npm start"? 
I saw many other sources, they can run with "npm start" without compiling scss files. Imagine that my folders have many new css files, it looks ugly. Thanks for any help!

kellyrmilligan replied:
Do you mean the initial terminal output? If so yes you can pass -q to build-css like so: 
`npm run build-css -- -q`, this will call it in quiet mode. if this is not why you mean, create react app does not let you ATM add in custom loaders for scss, so this is a workaround to allow for that. In practice it has t been a big deal tbh

kidqn replied:
Thank to your reply. I mean that I want to use "node-sass-chokidar" but still can import .scss files.
For example: my Home folder includes: Home.js, Home.scss, ... 
After run "npm start", my folder add Home.css. So I'm scared of that one day I forget and edit css files instead of scss.
Another issue is that it's assume that Styles folder has 10 SCSS files, and now after run "npm start" it will have 20 files. I knew we can ignore CSS files, but it makes structure of app folders cumbrous.

In some react apps built with webpack 2 as I did, we need just import .scss files, can run "npm start" in dev mode without compiling new css files. Therefore, can you improve it?

michaelwayman replied:
like @kellyrmilligan said, custom scss loaders are not supported at the moment.

 - One option you have available is to `npm run eject` and add a scss loader to `config/webpack.config.*.js`. *then* you should be able to `import 'styles.scss';`
 - Another option is to output the css to a separate folder with `node-sass-chokidar` with the `-o <path>` option, however, this will make importing the css a huge pain if you use self-contained component hierarchy pattern like I do.

I don't think the devs on this see this as highest priority, but perhaps you can convince them to let you open a PR providing an entry point or env variable into the web-pack-config WITHOUT running eject?

kidqn replied:
@michaelwayman : Thanks. That's exactly I want. 
I tried the second way, edit web-pack-config, run eject to add scss loader and eventually it failed.
The node-sass-chokidar's way works fine, so I think something likes scss loader will make node-sass-chokidar better :)

P/S: I'm using VS, and config it to hide css files.

Timer commented:
This is some great work, guys! Thanks so much.

FDiskas commented:
I suggest to use scripts like so
```
    "prebuild": "node-sass-chokidar src/ -o src/",
    "prestart": "npm run build-css && node-sass-chokidar src/ -o src/ --watch --recursive"
```
